### PR TITLE
fix semantic error

### PR DESCRIPTION
**Pull request by iattempt**
**Fixes issue:**
#3007 

**Changes:**
1. fix semantic error for decrease end before checker
2. remove functions for POD pointer/array, using iterator_traits to know value_type

### CODE REVIEW

File: code/utility/palindrome/palindrome_check/palindrome_check.cpp [hunk 1]
```diff
@@ -4,87 +4,55 @@
 
 namespace palindrome_check{
 template<typename _InputIter,
-         typename _ValueNotEqualTo = std::not_equal_to<typename _InputIter::value_type>,
+         typename _ValueNotEqualTo
+             = std::not_equal_to<typename std::iterator_traits<_InputIter>::value_type>,
```
Comment (iattempt) [inline code/utility/palindrome/palindrome_check/palindrome_check.cpp]: iterator_traits could extract value_type for STL and POD

File: code/utility/palindrome/palindrome_check/palindrome_check.cpp [hunk 2]
```diff
@@ -4,87 +4,55 @@
...
-
-        return isPalindromeRecursive<_POD>(++begin, end);
+        --end;
+        if (_IterLess()(begin, end))
+        {
+            if (_ValueNotEqualTo()(*begin, *end))
+                return false;
+
+            return isPalindromeRecursive<_InputIter, _ValueNotEqualTo, _IterLess>(++begin, end);
+        }
```
Comment (iattempt) [inline code/utility/palindrome/palindrome_check/palindrome_check.cpp]: before decrease end, check if is same node

**Comment by arnavb** [comment]
👍

**Comment by AdiChat** [comment]
Looks great 😍 

Hope you are enjoying your journey 🚆  with **OpenGenus**

**Pull request merged**